addons/bus/tests/test_assetsbundle.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from unittest.mock import patch
+
+import odoo.tests
+from odoo.osv import expression
+
+
+@odoo.tests.tagged('post_install', '-at_install', 'assets_bundle')
+class BusWebTests(odoo.tests.HttpCase):
+
+    def test_bundle_sends_bus(self):
+        """
+        Tests two things:
+        - Messages are post to the bus when assets change (their hash has been recomputed and differ from the attachment's)
+        - The interface deals with those bus messages by displaying one notification
+        """
+        # This is how many times longpolling is called before the manual re-enabling is reverted
+        # A too low limit may make the test crash because the patch may have been reverted too early
+        LONGPOLL_CALL_LIMIT = 5
+
+        db_name = self.env.registry.db_name
+        bundle_xml_ids = ('web.assets_common', 'web.assets_backend')
+
+        domain = []
+        for bundle in bundle_xml_ids:
+            domain = expression.OR([
+                domain,
+                [('name', 'ilike', bundle + '%')]
+            ])
+        # start from a clean slate
+        self.env['ir.attachment'].search(domain).unlink()
+        self.env.registry._clear_cache()
+
+        #
+        # Patch Methods
+        #
+
+        availability_patch = None
+        def stop_poll_availability_patch():
+            try:
+                availability_patch.stop()
+            except RuntimeError:
+                # If the patch is already stopped
+                pass
+
+        availability_calls = []
+        def patched_poll_check_availability():
+            """
+            The bus is not available in test mode, so we manually re-enable it
+            """
+            if len(availability_calls) >= LONGPOLL_CALL_LIMIT - 1:
+                stop_poll_availability_patch()
+            else:
+                availability_calls.append(True)
+
+        sendones = []
+        def patched_sendone(self, channel, message):
+            """
+            Control API and number of messages posted to the bus
+            """
+            sendones.append((channel, message))
+
+        _polls = []
+        def patched_poll(*args, **kwargs):
+            """
+            Force send a controlled response when longpolling asks
+            """
+            _polls.append('polled')
+            poll_id = len(_polls)
+            return [{
+                'id': poll_id,
+                'channel': (db_name, 'bundle_changed'),
+                'message': ('web.assets_backend', 'hash%s' % poll_id),
+            }]
+
+        #
+        # Enable Patches
+        #
+
+        availability_patch = patch('odoo.addons.bus.controllers.main.BusController._check_availability', wraps=patched_poll_check_availability)
```

Comment:
Any chance we could avoid this? Seems we could unit-test the bundle notification and UI reaction parts independently by:
- mocking the `sendone()` method to check that the notification is sent upon bundle change
- triggering a fake bus notification during the tour
- assuming the bus itself works - to avoid the horrors required to involve it during the test